tests/integration/test_charm.py

```diff
@@ -0,0 +1,125 @@
+    logger.info(f"Deploying {SLURMCTLD} against {SLURMD}, {SLURMDBD}, and {DATABASE}...")
+    slurmctld_res = get_slurmctld_res()
+    slurmd_res = get_slurmd_res()
+    await asyncio.gather(
+        ops_test.model.deploy(
+            str(await slurmctld_charm),
+            application_name=SLURMCTLD,
+            num_units=1,
+            resources=slurmctld_res,
+            series=series,
+        ),
+        ops_test.model.deploy(
+            SLURMD,
+            application_name=SLURMD,
+            channel="edge",
+            num_units=1,
+            resources=slurmd_res,
+            series=series,
+        ),
+        ops_test.model.deploy(
+            SLURMDBD, application_name=SLURMDBD, channel="edge", num_units=1, series=series
+        ),
+        ops_test.model.deploy(
+            DATABASE, application_name=DATABASE, channel="edge", num_units=1, series="bionic"
+        ),
+    )
+    # Attach resources to charms.
+    await ops_test.juju("attach-resource", SLURMCTLD, f"etcd={slurmctld_res['etcd']}")
+    await ops_test.juju("attach-resource", SLURMD, f"nhc={slurmd_res['nhc']}")
+    await ops_test.juju(
+        "attach-resource", SLURMD, f"singularity-deb={slurmd_res['singularity-deb']}"
+    )
+    await ops_test.juju(
+        "attach-resource", SLURMD, f"singularity-rpm={slurmd_res['singularity-rpm']}"
+    )
+    # Set integrations for charmed applications.
+    await ops_test.model.relate(f"{SLURMCTLD}:{SLURMD}", f"{SLURMD}:{SLURMD}")
+    await ops_test.model.relate(f"{SLURMCTLD}:{SLURMDBD}", f"{SLURMDBD}:{SLURMDBD}")
+    await ops_test.model.relate(f"{SLURMDBD}:db", f"{DATABASE}:db")
+    # Reduce the update status frequency to accelerate the triggering of deferred events.
+    async with ops_test.fast_forward():
+        await ops_test.model.wait_for_idle(apps=[SLURMCTLD], status="active", timeout=1000)
+        assert ops_test.model.applications[SLURMCTLD].units[0].workload_status == "active"
+
+
+@pytest.mark.abort_on_fail
+async def test_slurmctld_is_active(ops_test: OpsTest) -> None:
+    """Test that slurmctld is active inside Juju unit."""
+    logger.info("Checking that slurmctld is active inside Juju unit...")
+    async with unit_connection(ops_test, SLURMCTLD, UNIT) as conn:
+        stdin, stdout, stderr = conn.exec_command("systemctl is-active slurmctld")
+        assert codecs.decode(stdout.read()).strip("\n") == "active"
+
+
+@pytest.mark.abort_on_fail
+@tenacity.retry(
+    wait=tenacity.wait.wait_exponential(multiplier=2, min=1, max=30),
+    stop=tenacity.stop_after_attempt(10),
+    reraise=True,
+)
+async def test_slurmctld_port_listen(ops_test: OpsTest) -> None:
+    """Test that slurmctld is listening on port 6817."""
+    logger.info("Checking that slurmctld is listening on port 6817...")
+    async with unit_connection(ops_test, SLURMCTLD, UNIT) as conn:
+        conn.exec_command("sudo apt-get -y install net-tools")
```

Comment:
Hmm, can't you just use requests as seen here?: https://github.com/jnsgruk/parca-operator/blob/2b8a98a89101f67a7222d26ed72679cb7351c097/tests/integration/test_charm.py#L34

Reply:
Let me try that.

Reply:
Hmm... After some further investigation, I do not believe we can just use the _requests_ library instead of what I have here. Reason saying is that slurmctld does not use HTTP to communicate with the other SLURM services. `requests.get()` is a wrapper for issuing a GET command to an HTTP service. It wouldn't work for checking if the slurmctld port is active since SLURM just uses TCP.

Originally I did try to use `lsof` instead of `netstat`, but I could not get it to work the way I wanted to. `lsof` comes by default inside the Juju units, so let me see if I can get the command to work again. If so, I should be able to remove the install command for netstat

Reply:
Resolved in https://github.com/canonical/slurmctld-operator/pull/1/commits/8c5b6139462e9db40ec1096f09d8c0f8d7d02b94. Got the test down to four lines!